File: survey/domain/src/main/java/me/nalab/survey/domain/feedback/Reviewer.java
package me.nalab.survey.domain.feedback;

import java.util.function.LongSupplier;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import me.nalab.survey.domain.exception.IdAlreadyGeneratedException;
import me.nalab.survey.domain.support.IdGeneratable;

@Builder
@Getter
@EqualsAndHashCode
@ToString
public class Reviewer implements IdGeneratable {

	private Long id;
	private String nickName;
	private final boolean collaborationExperience;
	private final String position;

	public void generateFirstReviewerNickName() {
		nickName = "A";
	}

	public void generateNickName(String lastName) {
		nickName = getNextNickName(lastName);
	}

	private String getNextNickName(String lastName) {
		for(int i = lastName.length() - 1; i >= 0; i--) {
			if(lastName.charAt(i) != 'Z') {
				return lastName.substring(0, i) + (char)(lastName.charAt(i) + 1) + "A".repeat(lastName.length() - (i + 1));
			}
		}
		return "A".repeat(Math.max(0, lastName.length() + 1));
	}

	@Override
	public void withId(LongSupplier idSupplier) {
		if(this.id != null) {
			throw new IdAlreadyGeneratedException(this);
		}
		id = idSupplier.getAsLong();
	}

}

File: src/main/kotlin/me/nalab/api/survey/domain/feedback/Reviewer.kt
package me.nalab.api.survey.domain.feedback

import me.nalab.api.core.TimeBaseEntity
import me.nalab.api.survey.domain.feedback.value.NickName
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.Table

@Entity(name = "reviewer")
@Table(name = "reviewer")
class Reviewer(
    @Id
    @Column(name = "reviewer_id")
    val id: Long,

    @Column(name = "nick_name", nullable = false)
    internal var nickName: NickName,

    @Column(name = "collaboration_experience", nullable = false)
    val collaborationExperience: Boolean,

    @Column(name = "position", nullable = false)
    var position: String,
) : TimeBaseEntity()
